namespace/entry.go

```diff
@@ -0,0 +1,226 @@
+package namespace
+
+import (
+	"fmt"
+	"sort"
+	"strings"
+
+	"github.com/docker/distribution"
+)
+
+const (
+	actionPull      = "pull"
+	actionPush      = "push"
+	actionIndex     = "index"
+	actionNamespace = "namespace"
+)
+
+func checkAction(s string) error {
+	switch s {
+	case actionPull:
+		fallthrough
+	case actionPush:
+		fallthrough
+	case actionIndex:
+		fallthrough
+	case actionNamespace:
+		return nil
+	default:
+		return fmt.Errorf("action invalid: %q unsupported", s)
+	}
+}
+
+// Entry defines the contents of a namespace configuration.
+type Entry struct {
+	// scope identifies the namespaces to which the entry applies. If any name
+	// matches scope, the entry may apply.
+	scope scope
+
+	// action is an opaque action field.
+	action string
+
+	// args defines a set of arguments. The meaning and ordering of arguments
+	// are dependent on the particular action.
+	args []string
+}
+
+// NewEntry creates a new entry with the given scope and action.  If the scope,
+// action, or action arguments are invalid, this function will return an error.
+func NewEntry(scopeStr, action string, args ...string) (Entry, error) {
+	scp, err := parseScope(scopeStr)
+	if err != nil {
+		return Entry{}, fmt.Errorf("entry invalid: %v", err)
+	}
+
+	if err := checkAction(action); err != nil {
+		return Entry{}, fmt.Errorf("entry invalid: %v", err)
+	}
+	return Entry{
+		scope:  scp,
+		action: action,
+		args:   args,
+	}, nil
+}
+
+// Scope returns the scope of the entry.
+func (e Entry) Scope() distribution.Scope {
+	return e.scope
+}
+
+// Action returns the action of the entry.
+func (e Entry) Action() string {
+	return e.action
+}
+
+// Args return the arguments for the action of the entry.
+func (e Entry) Args() []string {
+	return e.args
+}
+
+// entryLess returns true if a is less than b, providing a total ordering for
+// entries for deterministically resolve and matching namespaces. The entries
+// are sorted lexically using scope, action, args, in that order.
+func entryLess(a, b Entry) bool {
+	if entryEqual(a, b) {
+		return false
+	}
+
+	// now, we start with scope and fallback to each field of the entry.
+	if a.scope != b.scope {
+		return a.scope < b.scope
+	}
+
+	// fallback to action
+	if a.action != b.action {
+		return a.action < b.action
+	}
+
+	return strings.Join(a.args, " ") < strings.Join(b.args, " ")
+}
+
+func entryEqual(a, b Entry) bool {
+	v := a.scope == b.scope &&
+		a.action == b.action &&
+		strings.Join(a.args, " ") == strings.Join(b.args, " ")
+
+	// log.Println(a, "==", b, v)
+	return v
+}
+
+// Entries are the primary mechanism under which namespaces are transported,
+// merged, stored and queried. This represents a unique set of entries.
+type Entries struct {
```

Comment:
This struct only exists to ensure the underlying entry array is sorted. This keeps inserts and lookups fast. We could get rid of this struct and use functions which mutate and return the Entry slice, but we will lose that guarantee.  This entries  slice is never actually sorted since inserts are in order.